geode-core/src/main/java/org/apache/geode/internal/cache/event/DistributedEventTracker.java

```diff
@@ -356,22 +363,58 @@ public boolean hasSeenEvent(EventID eventID, InternalCacheEvent tagHolder) {
       if (evh.isRemoved() || evh.getLastSequenceNumber() < eventID.getSequenceID()) {
         return false;
       }
-      // log at fine because partitioned regions can send event multiple times
-      // during normal operation during bucket region initialization
-      if (logger.isTraceEnabled(LogMarker.DISTRIBUTION_BRIDGE_SERVER_VERBOSE)) {
-        logger.trace(LogMarker.DISTRIBUTION_BRIDGE_SERVER_VERBOSE,
-            "Cache encountered replay of event with ID {}.  Highest recorded for this source is {}",
-            eventID, evh.getLastSequenceNumber());
+      if (shouldLogPreviouslySeenEvent(tagHolder, evh)) {
+        logger.info(EVENT_HAS_PREVIOUSLY_BEEN_SEEN, region.getName(),
+            tagHolder == null ? "unknown" : ((EntryEventImpl) tagHolder).getKey(),
+            tagHolder == null ? "unknown" : tagHolder.getOperation(), eventID.expensiveToString(),
+            evh.getLastSequenceNumber());
       }
       // bug #44956 - recover version tag for duplicate event
       if (evh.getLastSequenceNumber() == eventID.getSequenceID() && tagHolder != null
           && evh.getVersionTag() != null) {
         ((EntryEventImpl) tagHolder).setVersionTag(evh.getVersionTag());
       }
+
+      // Increment the previously seen events statistic
+      region.getCachePerfStats().incPreviouslySeenEvents();
+
       return true;
     }
   }
 
+  private boolean shouldLogPreviouslySeenEvent(InternalCacheEvent event,
+      EventSequenceNumberHolder evh) {
+    boolean shouldLogSeenEvent = true;
+    String message = null;
```

Comment:
@boglesby this method is ostensibly deciding whether or not to log the `event`. That being the case, I don't understand how it is possible (or sensible) for `event` to be `null`.

Reply:
This is a case in transactions. This is detailed in the method:
```java
  /**
   * has the Region's event state seen this event? Most checks should use the method that takes an
   * Event, not an ID, but with transactions we do not have an event at the time the check needs to
   * be made. Consequently, this method may cause events to be recorded that would otherwise be
   * ignored.
   *
   * @param eventID the identifier of the event
   * @return true if the Region's event state has seen the event
   */
  @Override
  public boolean hasSeenEvent(EventID eventID) {
    return getEventTracker().hasSeenEvent(eventID);
  }
```

Reply:
I didn't want to take a chance on an NPE since hasSeenEvent can be called with a null event in the case that Naba mentioned. One of the other cases that don't use the event posDup can still apply.